students/kristianjf/lesson10/mailroom_6.py

```diff
@@ -0,0 +1,302 @@
+        return sum(self.donations) < sum(other.donations)
+    @property
+    def name(self):
+        return self._name
+    @property
+    def donations(self):
+        return self._donations
+    @property
+    def normalized_name(self):
+        'Validate and Return Name Input'
+        full_name_cap = ''
+        try:
+            for name in self.name.split():
+                assert name.isalpha()
+                full_name_cap += f'{name.capitalize()} '
+            full_name = full_name_cap.strip()
+        except AssertionError:
+            print('Invalid Name. Found non-alphabetic characters')
+        else:
+            return full_name
+        return self._normalized_name
+    def donate(self, new_donation):
+        self.donations.append(new_donation)
+    def metrics(self):
+        num_d = len(self.donations)
+        total_d = sum(self.donations)
+        avg_d = total_d / num_d
+        return num_d, total_d, avg_d
+
+    def send_letter(self):
+        '''
+        Try to use a dict and the .format() method to do the letter as one big \
+        template rather than building up a big string in parts.
+        In this version, add a function (and a menu item to invoke it), that goes \
+        through all the donors in your donor data structure, generates a thank you \
+        letter, and writes it to disk as a text file.
+        '''
+        filename = '_'.join(self.name.split())+'.txt'
+        with open(filename, 'w') as outfile:
+            outfile.write(self.output_string().format(name=self.name, donation=self.donations))
+            print(f'Writing: {filename}')
+
+    def output_string(self):
+        'Output string for letters'
+        format_string = 'Dear {name},\n\nThank you for your generous donation of ' \
+                        '${donation}. Please send us more money at your earliest ' \
+                        'convenience.'
+        return format_string
+
+class DonorHandler:
+    def __init__(self):
+        self.donors = {}
+    def add_donor(self, donor_object):
+        self.donors[donor_object.normalized_name] = donor_object
+    def dnr_challenge(self, factor=1, **kwargs):
+        new_dh = DonorHandler()
+        for donor, donor_object in self.donors.items():
+            # new_d = Donor(donor, list(map(lambda x, y: x*y, \
+            # donor_object.donations, len(donor_object.donations)*[factor])))
+            # new_d = Donor(donor, \
+            # self.dnr_filter(\
+            # list(map(lambda x, y: x*y, donor_object.donations, \
+            # len(donor_object.donations)*[factor])),\
+            # kwargs['min_donation'], kwargs['max_donation']))
+            # new_dh.add_donor(new_d)
+            '''Add a new feature to Mailroom using filter so that donations either above
+            or below a specified dollar amount are included in the map operations of #1 above.
+            You can do this by adding min_donation and max_donation optional keyword parameters
```

Comment:
this means something like:

`def challenge( min_donation=None, max_donation=None):`

that's more clear what you need than simply expecting it to be in kwargs....